In [1]:
import pandas as pd

In [2]:
df = pd.read_csv("model_building.csv")
df.head()

title_x  average_rating   
0  instant compostable espresso capsules lungo me...             4.3  \
1  instant compostable espresso capsules lungo me...             4.3   
2  instant compostable espresso capsules lungo me...             4.3   
3  edible markersfood coloring markersfood colori...             4.3   
4  edible markersfood coloring markersfood colori...             4.3   

   rating_number  price parent_asin  rating   
0             85   8.49  B0C2W77WJX       4  \
1             85   8.49  B0C2W77WJX       5   
2             85   8.49  B0C2W77WJX       4   
3           1193   8.99  B07PK9L29R       5   
4           1193   8.99  B07PK9L29R       5   

                                             title_y   
0  fresh tasting smelling slightly acidic light l...  \
1  dynamic flavor interesting flavor profile body...   
2                                 pricey much flavor   
3                                                fun   
4                                      perfect touch   

                                                text   
0  happen instant pod dual coffee maker spots nes...  \
1  tried leggero light roast lungo medium roast d...   
2  great roast ppl arent bitter heavy taste like ...   
3  much fun color cookies artist really work well...   
4  perfect adding creative touches taste coloring...   

                        user_id  category_Grocery  
0  AF2BLE54TEMGZ546U763ZHZRXC4A                 1  
1  AF2BLE54TEMGZ546U763ZHZRXC4A                 1  
2  AEUDZQDVSZYCHEXQSXLB6NWQTMHA                 1  
3  AGECC4F4CDL2AVODIRNCF3V63BEQ                 1  
4  AFF6LERKD46F2RLIKAMQTAQPOIWA                 1

In [25]:
print(df.shape)

(110596, 10)


## Data Preprocessing: Tokenization
### Objective: Convert text data into tokens for further analysis.

#### Steps:

- Import Libraries: Utilize nltk for natural language processing tasks.
- Download Necessary Resources: Ensure the ‘punkt’ tokenizer models are available.
- Data Conversion: Cast the ‘text’ column to string type to avoid type-related errors.
- Tokenization: Apply word_tokenize from nltk to split the text into individual words or tokens.
- Store Tokens: Save the tokenized data in a new column ‘tokenized_reviews’ for easy access.

In [3]:
# Ensure nltk is downloaded and imported
import nltk
nltk.download('punkt')
from nltk.tokenize import word_tokenize

# Convert any non-string data to string
df['text'] = df['text'].astype(str)

# Tokenize the text data
tokenized_data = df['text'].apply(lambda x: word_tokenize(x))


df['tokenized_reviews'] = tokenized_data

df.head()

[nltk_data] Downloading package punkt to
[nltk_data]     C:\Users\Hrishikesh\AppData\Roaming\nltk_data...
[nltk_data]   Package punkt is already up-to-date!


title_x  average_rating   
0  instant compostable espresso capsules lungo me...             4.3  \
1  instant compostable espresso capsules lungo me...             4.3   
2  instant compostable espresso capsules lungo me...             4.3   
3  edible markersfood coloring markersfood colori...             4.3   
4  edible markersfood coloring markersfood colori...             4.3   

   rating_number  price parent_asin  rating   
0             85   8.49  B0C2W77WJX       4  \
1             85   8.49  B0C2W77WJX       5   
2             85   8.49  B0C2W77WJX       4   
3           1193   8.99  B07PK9L29R       5   
4           1193   8.99  B07PK9L29R       5   

                                             title_y   
0  fresh tasting smelling slightly acidic light l...  \
1  dynamic flavor interesting flavor profile body...   
2                                 pricey much flavor   
3                                                fun   
4                                      perfect touch   

                                                text   
0  happen instant pod dual coffee maker spots nes...  \
1  tried leggero light roast lungo medium roast d...   
2  great roast ppl arent bitter heavy taste like ...   
3  much fun color cookies artist really work well...   
4  perfect adding creative touches taste coloring...   

                        user_id  category_Grocery   
0  AF2BLE54TEMGZ546U763ZHZRXC4A                 1  \
1  AF2BLE54TEMGZ546U763ZHZRXC4A                 1   
2  AEUDZQDVSZYCHEXQSXLB6NWQTMHA                 1   
3  AGECC4F4CDL2AVODIRNCF3V63BEQ                 1   
4  AFF6LERKD46F2RLIKAMQTAQPOIWA                 1   

                                   tokenized_reviews  
0  [happen, instant, pod, dual, coffee, maker, sp...  
1  [tried, leggero, light, roast, lungo, medium, ...  
2  [great, roast, ppl, arent, bitter, heavy, tast...  
3  [much, fun, color, cookies, artist, really, wo...  
4  [perfect, adding, creative, touches, taste, co...

In [4]:
# now as the data (text - reviews data) is tokenised it's time to drop it the main 'text' column
df = df.drop('text', axis = 1)
df.head()

title_x  average_rating   
0  instant compostable espresso capsules lungo me...             4.3  \
1  instant compostable espresso capsules lungo me...             4.3   
2  instant compostable espresso capsules lungo me...             4.3   
3  edible markersfood coloring markersfood colori...             4.3   
4  edible markersfood coloring markersfood colori...             4.3   

   rating_number  price parent_asin  rating   
0             85   8.49  B0C2W77WJX       4  \
1             85   8.49  B0C2W77WJX       5   
2             85   8.49  B0C2W77WJX       4   
3           1193   8.99  B07PK9L29R       5   
4           1193   8.99  B07PK9L29R       5   

                                             title_y   
0  fresh tasting smelling slightly acidic light l...  \
1  dynamic flavor interesting flavor profile body...   
2                                 pricey much flavor   
3                                                fun   
4                                      perfect touch   

                        user_id  category_Grocery   
0  AF2BLE54TEMGZ546U763ZHZRXC4A                 1  \
1  AF2BLE54TEMGZ546U763ZHZRXC4A                 1   
2  AEUDZQDVSZYCHEXQSXLB6NWQTMHA                 1   
3  AGECC4F4CDL2AVODIRNCF3V63BEQ                 1   
4  AFF6LERKD46F2RLIKAMQTAQPOIWA                 1   

                                   tokenized_reviews  
0  [happen, instant, pod, dual, coffee, maker, sp...  
1  [tried, leggero, light, roast, lungo, medium, ...  
2  [great, roast, ppl, arent, bitter, heavy, tast...  
3  [much, fun, color, cookies, artist, really, wo...  
4  [perfect, adding, creative, touches, taste, co...

## TF-IDF Vectorization

#### To convert the tokenized reviews into a numerical format that can be used for machine learning models, we use the `TfidfVectorizer` from `sklearn.feature_extraction.text`. This process is known as TF-IDF vectorization.

### Code Explanation

- **Tokenized Reviews**: We start with the `tokenized_reviews` column which contains the tokenized text data.
- **String Conversion**: Each list of tokens is joined into a single string.
- **Hyperparameters**: We define hyperparameters such as `ngram_range` to include unigrams and bigrams, and `min_df` to set the minimum document frequency for terms.
- **TF-IDF Vectorizer Initialization**: The vectorizer is initialized with the defined hyperparameters.
- **Fitting and Transforming**: The vectorizer is then fitted to the tokenized strings and transforms them into a TF-IDF matrix.
- **Matrix Shape**: Finally, we print the shape of the TF-IDF matrix to understand the dimensions of our feature space.

In [5]:
from sklearn.feature_extraction.text import TfidfVectorizer

# Assuming 'tokenized_reviews' column contains the tokenized reviews
tokenized_reviews = df['tokenized_reviews']

# Convert tokenized reviews into strings
tokenized_reviews_str = tokenized_reviews.apply(lambda x: ' '.join(x))

# # Define hyperparameters
# ngram_range = (1, 2)  # Include unigrams and bigrams
# min_df = 2  # Ignore terms that have a document frequency strictly lower than the given threshold

# Initialize TF-IDF vectorizer with hyperparameters
tfidf_vectorizer = TfidfVectorizer(min_df=5, max_features=10000, strip_accents='unicode', 
                                   analyzer='word', ngram_range=(1, 3), stop_words='english')

# Fit and transform the tokenized reviews
tfidf_matrix = tfidf_vectorizer.fit_transform(tokenized_reviews_str)

# Print the shape of the TF-IDF matrix
print("Shape of TF-IDF matrix:", tfidf_matrix.shape)


Shape of TF-IDF matrix: (110596, 10000)


### Here’s a breakdown of what these vectors represent:

- TF-IDF: Stands for Term Frequency-Inverse Document Frequency. It’s a numerical statistic that is intended to reflect how important a word is to a document in a collection or corpus.
- Vector Format: Each vector is in the format (document_index, word_id) tf-idf_score.
- document_index: The index of the document in the corpus.
- word_id: The unique identifier for a word within the corpus.
- tf-idf_score: The TF-IDF score for that word in the specified document.
- Purpose: These vectors are used to convert text data into a numerical form that can be used for various tasks such as clustering, classification, and information retrieval.

#### For example, the vector (0, 8688) 0.13354025856957072 means that in the first document of the corpus (index 0), the word with ID 8688 has a TF-IDF score of approximately 0.134. This score represents the relative importance of this word in that particular document compared to the entire corpus. High scores indicate more importance, and vice versa. The vectors are typically sparse, meaning most of the values are zero, as a word does not appear in most documents.

In [6]:
print(tfidf_matrix)

  (0, 8688)	0.13354025856957072
  (0, 1766)	0.1477192016466634
  (0, 8231)	0.14581725775522944
  (0, 3729)	0.1422139670778428
  (0, 4907)	0.12113575777422866
  (0, 5586)	0.11079168371886787
  (0, 5033)	0.14674024378153716
  (0, 8940)	0.13783991642199725
  (0, 1007)	0.1425768555929223
  (0, 3198)	0.09303436837432928
  (0, 5851)	0.14722230523111116
  (0, 1751)	0.11870223660923207
  (0, 8108)	0.15665448279818145
  (0, 1166)	0.1067334936829738
  (0, 6713)	0.1072603212643369
  (0, 4073)	0.09238200984060295
  (0, 8675)	0.06041627082806357
  (0, 3439)	0.129431934122404
  (0, 7855)	0.08894018262606979
  (0, 1817)	0.08011118611107762
  (0, 6423)	0.11498223399854929
  (0, 7376)	0.1185357623581828
  (0, 521)	0.10409122876522635
  (0, 5821)	0.09728635957992757
  (0, 899)	0.10272855302536375
  :	:
  (110593, 9382)	0.1297708145442301
  (110593, 6122)	0.16092951754191354
  (110593, 7015)	0.1272452582061893
  (110593, 6636)	0.13486952244804257
  (110593, 8514)	0.09444501276805845
  (110593, 4916)	0.17

### Sparse Matrix Conversion
The code snippet is converting a dense TF-IDF matrix to a sparse matrix format using the `csr_matrix` class from the `scipy.sparse` module. This is done to optimize memory usage when dealing with large matrices.

### Output Explanation
- **Type of tfidf_matrix_sparse**: Confirms the matrix is now in CSR format.
- **Shape of tfidf_matrix_sparse**: The matrix dimensions reflect the number of documents (rows) and the maximum number of features (columns) considered in the TF-IDF vectorization.
- **Number of non-zero entries**: Indicates the total count of non-zero values in the matrix, which represents the actual data points that are stored in memory.


In [7]:
from scipy.sparse import csr_matrix

# Convert to a sparse matrix (if not already)
tfidf_matrix_sparse = csr_matrix(tfidf_matrix)

# Print the type and shape of the sparse TF-IDF matrix
print("Type of tfidf_matrix_sparse:", type(tfidf_matrix_sparse))
print("Shape of tfidf_matrix_sparse:", tfidf_matrix_sparse.shape)
print("Number of non-zero entries:", tfidf_matrix_sparse.nnz)

Type of tfidf_matrix_sparse: <class 'scipy.sparse._csr.csr_matrix'>
Shape of tfidf_matrix_sparse: (110596, 10000)
Number of non-zero entries: 2118646


since i was encountering this error - Unable to allocate 52.8 GiB for an array with shape (110596, 64024) and data type float64
i'm using gensim library,  Gensim has an efficient TF-IDF model that doesn’t require loading the entire corpus into memory at once. Instead, it operates on an iterable representation of your data.

Here’s how you can use Gensim for TF-IDF representation:

Install Gensim (if you haven’t already):
pip install gensim

Prepare Your Corpus:
Ensure your corpus is an iterable (e.g., a list of tokenized reviews).
You can use Python generators to achieve this.
Create a Gensim Dictionary:
Create a dictionary from your tokenized reviews.
The dictionary maps words to unique integer IDs.
Create a Gensim Corpus:
Convert your tokenized reviews into a Gensim corpus using the dictionary.
The corpus is a list of bag-of-words representations (sparse vectors) for each document.
Compute TF-IDF:
Use Gensim’s TfidfModel to compute the TF-IDF scores based on the corpus.
Access the TF-IDF Vectors:
You can access the TF-IDF vectors for individual documents without loading the entire matrix into memory.

### Thoughts behind using gensim and corpora - 
- Importing Gensim: We start by importing Gensim, a versatile Python library renowned for its capabilities in natural language processing tasks such as topic modeling, document similarity analysis, and text summarization. This step ensures we have access to the tools necessary for our task ahead.
- Creating a Dictionary: In this step, we construct a dictionary that serves as a fundamental component in the subsequent processes. This dictionary acts as a mapping between words and their unique integer IDs within our corpus of product reviews. By establishing this mapping, we lay the groundwork for efficient representation and analysis of textual data.
- Creating a Corpus: With our dictionary in place, we proceed to create a corpus, essentially a collection of documents represented in a format suitable for computational analysis. Each document is transformed into a bag-of-words vector, where each word's frequency within the document is recorded. This step transforms our raw textual data into a structured format amenable to further processing.
- TF-IDF Model: Building upon the bag-of-words representation, we leverage Gensim's capabilities to construct a TF-IDF model. TF-IDF (Term Frequency-Inverse Document Frequency) is a statistical measure used to evaluate the importance of a word in a document relative to a corpus. By applying this model to our corpus, we compute TF-IDF scores for each word, capturing their significance within individual documents and across the entire corpus.
- Accessing TF-IDF Vectors: Here, we access the TF-IDF vectors generated by our model, enabling us to explore and utilize the transformed textual data. These vectors encapsulate the essence of each document's content, with each dimension representing a unique word and its corresponding TF-IDF score. By retrieving these vectors, we gain valuable insights into the semantic composition of our product reviews, paving the way for various downstream applications.

In [8]:
import gensim
from gensim import corpora, models

# Assuming 'tokenized_reviews' is an iterable (e.g., list of lists)
dictionary = corpora.Dictionary(tokenized_reviews)
corpus = [dictionary.doc2bow(review) for review in tokenized_reviews]

# Create the TF-IDF model
tfidf_model = models.TfidfModel(corpus)

# # Get the TF-IDF vectors for a specific document (e.g., first document)
# doc_index = 0
# tfidf_vector = tfidf_model[corpus[doc_index]]

# # Print the TF-IDF vector for the first document
# print(tfidf_vector)


# Create a list to hold all TF-IDF vectors
all_tfidf_vectors = []

# Iterate over all documents in the corpus
for doc in corpus:
    # Apply the TF-IDF model to get the vector for the current document
    doc_tfidf_vector = tfidf_model[doc]
    # Append the vector to the list
    all_tfidf_vectors.append(doc_tfidf_vector)

# Now 'all_tfidf_vectors' contains the TF-IDF vectors for all documents
# Limit the output to the first 10 TF-IDF vectors
limited_tfidf_vectors = all_tfidf_vectors[:10]

# Now 'limited_tfidf_vectors' contains the TF-IDF vectors for the first 10 documents
print(limited_tfidf_vectors)

[[(0, 0.15822019336594337), (1, 0.07911009668297168), (2, 0.12170821532633551), (3, 0.07493418887314095), (4, 0.11030375926386878), (5, 0.10859182482154411), (6, 0.04947903549083479), (7, 0.10374842137816317), (8, 0.1136241355227434), (9, 0.13596692814909372), (10, 0.13138885661596642), (11, 0.09599582425095556), (12, 0.1565601201334325), (13, 0.08020523861709881), (14, 0.08266694420118104), (15, 0.11723536349072858), (16, 0.18658037825369134), (17, 0.0888713835981412), (18, 0.07936684071231864), (19, 0.07183341245268707), (20, 0.12707538562583412), (21, 0.07401393015048931), (22, 0.032219935791116754), (23, 0.12179428611948441), (24, 0.05987914001758319), (25, 0.120831002481178), (26, 0.09560107520762093), (27, 0.06796589385054766), (28, 0.10209478190637414), (29, 0.11026127062439335), (30, 0.2560739659343325), (31, 0.028176078444046855), (32, 0.11693806256431222), (33, 0.1019816764207001), (34, 0.11392707128848491), (35, 0.1524549702782134), (36, 0.3968735655113111), (37, 0.101757343

In [9]:
print(len(all_tfidf_vectors))

110596


#### Applying padding on the vectors to obtain it in equal lengths 
- Importing NumPy: The code begins by importing the NumPy library, which is a fundamental package for scientific computing in Python.
- Max Length Calculation: max_length is determined by finding the longest vector in all_tfidf_vectors, which is a list of TF-IDF vectors. This ensures that all vectors will be padded to the same length.
- Padding Function: pad_vector is a function that takes a vector and the max_length as arguments. It calculates the necessary padding (a list of zeros) to make the vector’s length equal to max_length. The padding is then appended to the original vector, and the padded vector is returned.
- Applying Padding: The list comprehension applies the pad_vector function to each vector in all_tfidf_vectors. The result is padded_vectors, a new list where each TF-IDF vector has been padded with zeros to have the same length, ensuring uniformity for further processing.

In [10]:
import numpy as np

# Assuming 'all_tfidf_vectors' is a list of TF-IDF vectors
max_length = max(len(vector) for vector in all_tfidf_vectors)

# Function to pad the vectors
def pad_vector(vector, max_length):
    padding = [0] * (max_length - len(vector))
    return vector + padding

# Apply padding to each vector
padded_vectors = [pad_vector(vector, max_length) for vector in all_tfidf_vectors]


In [11]:
# Assuming 'padded_vectors' is a list of lists (vectors)
vector_lengths = [len(vector) for vector in padded_vectors]

# Check if all vectors are of the same length
if len(set(vector_lengths)) == 1:
    print("All vectors are of the same length:", vector_lengths[0])
else:
    print("Vectors are of varying lengths")
limited_padded_vectors = padded_vectors[:10]

# Now 'limited_tfidf_vectors' contains the TF-IDF vectors for the first 10 documents
print(limited_padded_vectors)

All vectors are of the same length: 576
[[(0, 0.15822019336594337), (1, 0.07911009668297168), (2, 0.12170821532633551), (3, 0.07493418887314095), (4, 0.11030375926386878), (5, 0.10859182482154411), (6, 0.04947903549083479), (7, 0.10374842137816317), (8, 0.1136241355227434), (9, 0.13596692814909372), (10, 0.13138885661596642), (11, 0.09599582425095556), (12, 0.1565601201334325), (13, 0.08020523861709881), (14, 0.08266694420118104), (15, 0.11723536349072858), (16, 0.18658037825369134), (17, 0.0888713835981412), (18, 0.07936684071231864), (19, 0.07183341245268707), (20, 0.12707538562583412), (21, 0.07401393015048931), (22, 0.032219935791116754), (23, 0.12179428611948441), (24, 0.05987914001758319), (25, 0.120831002481178), (26, 0.09560107520762093), (27, 0.06796589385054766), (28, 0.10209478190637414), (29, 0.11026127062439335), (30, 0.2560739659343325), (31, 0.028176078444046855), (32, 0.11693806256431222), (33, 0.1019816764207001), (34, 0.11392707128848491), (35, 0.1524549702782134), (3

### Model Building and its steps
- Normalization: The TF-IDF matrix, stored in tfidf_matrix_sparse, is normalized using sklearn's normalize function. Normalization ensures that each TF-IDF vector has a unit norm, which can improve the performance of subsequent operations.
- Dimensionality Reduction: Truncated Singular Value Decomposition (SVD) is applied to the normalized TF-IDF matrix to reduce its dimensionality while preserving important information. This step is crucial for managing computational complexity and capturing the most relevant features.
- Chunk-based Similarity Calculation: To handle large datasets efficiently, the code defines a function chunk_similarity that calculates cosine similarity scores between a reference document and chunks of the TF-IDF matrix. This approach enables memory-efficient computation by processing the data in manageable chunks.
- Recommendation Generation: Another function, get_recommendations_in_chunks, utilizes the chunk-based similarity calculation to generate recommendations for a given product index. It calculates cosine similarity scores between the reference product and all other products in the dataset, filtering out low similarity scores based on a threshold. Finally, it returns the top N products with the highest similarity scores as recommendations.

In [12]:
from sklearn.metrics.pairwise import cosine_similarity
from sklearn.preprocessing import normalize
from sklearn.decomposition import TruncatedSVD
import numpy as np

# Normalize the TF-IDF matrix
normalized_tfidf_matrix_sparse = normalize(tfidf_matrix_sparse)

# Initialize Truncated SVD with desired number of components
n_components = 100  # Adjust the number of components as needed
svd = TruncatedSVD(n_components=n_components)

# Apply Truncated SVD to the normalized TF-IDF matrix
tfidf_matrix_reduced = svd.fit_transform(normalized_tfidf_matrix_sparse)

# Define a function to calculate cosine similarity in chunks
def chunk_similarity(matrix, ref_idx, chunk_size=1000):
    num_chunks = matrix.shape[0] // chunk_size + (1 if matrix.shape[0] % chunk_size != 0 else 0)
    sim_scores = []

    for chunk_start in range(0, matrix.shape[0], chunk_size):
        chunk_end = min(chunk_start + chunk_size, matrix.shape[0])
        sim_chunk = cosine_similarity(matrix[ref_idx:ref_idx+1], matrix[chunk_start:chunk_end])
        sim_scores.extend(sim_chunk.flatten())

    return sim_scores

# Get recommendations based on cosine similarity matrix in chunks
def get_recommendations_in_chunks(product_index, matrix, n=5, threshold=0.2, chunk_size=1000):
    num_docs = matrix.shape[0]
    sim_scores = np.zeros(num_docs)

    for chunk_start in range(0, num_docs, chunk_size):
        chunk_end = min(chunk_start + chunk_size, num_docs)
        sim_chunk = cosine_similarity(matrix[product_index:product_index+1], matrix[chunk_start:chunk_end])
        sim_scores[chunk_start:chunk_end] = sim_chunk.flatten()

    # Filter out low similarity scores based on threshold
    sim_indices = [(idx, score) for idx, score in enumerate(sim_scores) if score > threshold]
    # Sort the products based on similarity scores
    sim_indices = sorted(sim_indices, key=lambda x: x[1], reverse=True)
    # Get the top similar products
    top_similar_products = sim_indices[:n]  # Limit to top N
    return top_similar_products

### Getting Recommendations

In [13]:
from tabulate import tabulate

# Example usage
product_index = 0
top_similar_products = get_recommendations_in_chunks(product_index, tfidf_matrix_reduced)

# Prepare the data for tabular display
table_data = []
for index, score in top_similar_products:
    title = df.iloc[index]['title_x']
    price = df.iloc[index]['price']
    avg_rating = df.iloc[index]['average_rating']
    table_data.append([title, price, avg_rating, score])

# Display the recommendations in a tabular format
print("Top 5 recommended products:")
print(tabulate(table_data, headers=['Title', 'Price', 'Avg Rating', 'Cosine Similarity Score'], tablefmt='pretty'))


Top 5 recommended products:
+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------+-------+------------+-------------------------+
|                                                                                 Title                                                                                 | Price | Avg Rating | Cosine Similarity Score |
+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------+-------+------------+-------------------------+
| instant compostable espresso capsules lungo medium roast 10 plantbased capsules makers instant pot ecofriendly 100 organic arabica capsules compostable freshness bag | 8.49  |    4.3     |   1.0000000000000004    |
|                   holland valley coffee keurig kcup coffee maker compatible high caffeine roast 100 or

### Evaluation

In [14]:
import pandas as pd
from sklearn.metrics import ndcg_score
import numpy as np

# Assuming get_recommendations_in_chunks is defined and returns a list of tuples (index, score)
product_index = 0
top_similar_products = get_recommendations_in_chunks(product_index, tfidf_matrix_reduced)

# Extract the indices and scores from the recommendations
top_indices = [index for index, score in top_similar_products]
top_scores = [score for index, score in top_similar_products]

# Assuming 'df' is your DataFrame and 'average_rating' is a column in your dataset
# Prepare relevance scores based on the average_rating of the recommended products
relevance_scores = df.loc[top_indices, 'average_rating'].tolist()

# Normalize relevance scores since NDCG in sklearn assumes relevance scores, not ratings directly
max_rating = max(relevance_scores)
normalized_relevance_scores = [score / max_rating for score in relevance_scores]

# Reshape for ndcg_score function
true_relevance = np.asarray([normalized_relevance_scores])
predicted_relevance = np.asarray([top_scores])

# Calculate NDCG score
ndcg = ndcg_score(true_relevance, predicted_relevance)
print("NDCG Score:", ndcg)


NDCG Score: 0.9967362727776101


### Using bi gram
#### For this model, all the above pre-processing and feature engineering were performed similarly below for bigram.

In [15]:
# Initialize TF-IDF vectorizer to use only bigrams
tfidf_vectorizer_bigram = TfidfVectorizer(min_df=5, max_features=10000, strip_accents='unicode', 
                                   analyzer='word', ngram_range=(2, 2), stop_words='english')

# Fit and transform the tokenized reviews
tfidf_matrix_bigram = tfidf_vectorizer_bigram.fit_transform(tokenized_reviews_str)

# Print the shape of the TF-IDF matrix
print("Shape of TF-IDF matrix:", tfidf_matrix_bigram.shape)

Shape of TF-IDF matrix: (110596, 10000)


In [16]:


# Convert to a sparse matrix (if not already)
tfidf_matrix_sparse_bigram = csr_matrix(tfidf_matrix_bigram)

# Print the type and shape of the sparse TF-IDF matrix
print("Type of tfidf_matrix_sparse_bigram:", type(tfidf_matrix_sparse_bigram))
print("Shape of tfidf_matrix_sparse_bigram:", tfidf_matrix_sparse_bigram.shape)
print("Number of non-zero entries:", tfidf_matrix_sparse_bigram.nnz)

Type of tfidf_matrix_sparse_bigram: <class 'scipy.sparse._csr.csr_matrix'>
Shape of tfidf_matrix_sparse_bigram: (110596, 10000)
Number of non-zero entries: 490379


In [17]:

# Assuming 'tokenized_reviews' is an iterable (e.g., list of lists)
dictionary_bigram = corpora.Dictionary(tokenized_reviews)
corpus_bigram = [dictionary_bigram.doc2bow(review) for review in tokenized_reviews]

# Create the TF-IDF model
tfidf_model_bigram = models.TfidfModel(corpus_bigram)

# Create a list to hold all TF-IDF vectors
all_tfidf_vectors_bigram = []

# Iterate over all documents in the corpus
for doc in corpus_bigram:
    # Apply the TF-IDF model to get the vector for the current document
    doc_tfidf_vector = tfidf_model[doc]
    # Append the vector to the list
    all_tfidf_vectors_bigram.append(doc_tfidf_vector)

# Now 'all_tfidf_vectors' contains the TF-IDF vectors for all documents
# Limit the output to the first 10 TF-IDF vectors
limited_tfidf_vectors_bigram = all_tfidf_vectors_bigram[:10]

# Now 'limited_tfidf_vectors' contains the TF-IDF vectors for the first 10 documents
print(limited_tfidf_vectors_bigram)

[[(0, 0.15822019336594337), (1, 0.07911009668297168), (2, 0.12170821532633551), (3, 0.07493418887314095), (4, 0.11030375926386878), (5, 0.10859182482154411), (6, 0.04947903549083479), (7, 0.10374842137816317), (8, 0.1136241355227434), (9, 0.13596692814909372), (10, 0.13138885661596642), (11, 0.09599582425095556), (12, 0.1565601201334325), (13, 0.08020523861709881), (14, 0.08266694420118104), (15, 0.11723536349072858), (16, 0.18658037825369134), (17, 0.0888713835981412), (18, 0.07936684071231864), (19, 0.07183341245268707), (20, 0.12707538562583412), (21, 0.07401393015048931), (22, 0.032219935791116754), (23, 0.12179428611948441), (24, 0.05987914001758319), (25, 0.120831002481178), (26, 0.09560107520762093), (27, 0.06796589385054766), (28, 0.10209478190637414), (29, 0.11026127062439335), (30, 0.2560739659343325), (31, 0.028176078444046855), (32, 0.11693806256431222), (33, 0.1019816764207001), (34, 0.11392707128848491), (35, 0.1524549702782134), (36, 0.3968735655113111), (37, 0.101757343

In [18]:
# Assuming 'all_tfidf_vectors' is a list of TF-IDF vectors
max_length = max(len(vector) for vector in all_tfidf_vectors_bigram)

# Function to pad the vectors
def pad_vector_bigram(vector, max_length):
    padding = [0] * (max_length - len(vector))
    return vector + padding

# Apply padding to each vector
padded_vectors_bigram = [pad_vector_bigram(vector, max_length) for vector in all_tfidf_vectors_bigram]


In [19]:
# Assuming 'padded_vectors' is a list of lists (vectors)
vector_lengths = [len(vector) for vector in padded_vectors_bigram]

# Check if all vectors are of the same length
if len(set(vector_lengths)) == 1:
    print("All vectors are of the same length:", vector_lengths[0])
else:
    print("Vectors are of varying lengths")
limited_padded_vectors_bigram = padded_vectors_bigram[:10]

# Now 'limited_tfidf_vectors' contains the TF-IDF vectors for the first 10 documents
print(limited_padded_vectors_bigram)

All vectors are of the same length: 576
[[(0, 0.15822019336594337), (1, 0.07911009668297168), (2, 0.12170821532633551), (3, 0.07493418887314095), (4, 0.11030375926386878), (5, 0.10859182482154411), (6, 0.04947903549083479), (7, 0.10374842137816317), (8, 0.1136241355227434), (9, 0.13596692814909372), (10, 0.13138885661596642), (11, 0.09599582425095556), (12, 0.1565601201334325), (13, 0.08020523861709881), (14, 0.08266694420118104), (15, 0.11723536349072858), (16, 0.18658037825369134), (17, 0.0888713835981412), (18, 0.07936684071231864), (19, 0.07183341245268707), (20, 0.12707538562583412), (21, 0.07401393015048931), (22, 0.032219935791116754), (23, 0.12179428611948441), (24, 0.05987914001758319), (25, 0.120831002481178), (26, 0.09560107520762093), (27, 0.06796589385054766), (28, 0.10209478190637414), (29, 0.11026127062439335), (30, 0.2560739659343325), (31, 0.028176078444046855), (32, 0.11693806256431222), (33, 0.1019816764207001), (34, 0.11392707128848491), (35, 0.1524549702782134), (3

In [20]:
# Normalize the TF-IDF matrix
normalized_tfidf_matrix_sparse_bigram = normalize(tfidf_matrix_sparse_bigram)

# Initialize Truncated SVD with desired number of components
n_components = 100  # Adjust the number of components as needed
svd = TruncatedSVD(n_components=n_components)

# Apply Truncated SVD to the normalized TF-IDF matrix
tfidf_matrix_reduced_bigram = svd.fit_transform(normalized_tfidf_matrix_sparse_bigram)

# Define a function to calculate cosine similarity in chunks
def chunk_similarity(matrix, ref_idx, chunk_size=1000):
    num_chunks = matrix.shape[0] // chunk_size + (1 if matrix.shape[0] % chunk_size != 0 else 0)
    sim_scores = []

    for chunk_start in range(0, matrix.shape[0], chunk_size):
        chunk_end = min(chunk_start + chunk_size, matrix.shape[0])
        sim_chunk = cosine_similarity(matrix[ref_idx:ref_idx+1], matrix[chunk_start:chunk_end])
        sim_scores.extend(sim_chunk.flatten())

    return sim_scores

# Get recommendations based on cosine similarity matrix in chunks
def get_recommendations_in_chunks(product_index, matrix, n=5, threshold=0.2, chunk_size=1000):
    num_docs = matrix.shape[0]
    sim_scores = np.zeros(num_docs)

    for chunk_start in range(0, num_docs, chunk_size):
        chunk_end = min(chunk_start + chunk_size, num_docs)
        sim_chunk = cosine_similarity(matrix[product_index:product_index+1], matrix[chunk_start:chunk_end])
        sim_scores[chunk_start:chunk_end] = sim_chunk.flatten()

    # Filter out low similarity scores based on threshold
    sim_indices = [(idx, score) for idx, score in enumerate(sim_scores) if score > threshold]
    # Sort the products based on similarity scores
    sim_indices = sorted(sim_indices, key=lambda x: x[1], reverse=True)
    # Get the top similar products
    top_similar_products_bigram = sim_indices[:n]  # Limit to top N
    return top_similar_products_bigram



### Getting Recommendations for Bigram

In [37]:
# Example usage
product_index = 45
top_similar_products_bigram = get_recommendations_in_chunks(product_index, tfidf_matrix_reduced_bigram)

# Prepare the data for tabular display
table_data = []
for index, score in top_similar_products_bigram:
    title = df.iloc[index]['title_x']
    price = df.iloc[index]['price']
    avg_rating = df.iloc[index]['average_rating']
    table_data.append([title, price, avg_rating, score])

# Display the recommendations in a tabular format
print("Top 5 recommended products:")
print(tabulate(table_data, headers=['Title', 'Price', 'Avg Rating', 'Cosine Similarity Score'], tablefmt='pretty'))


Top 5 recommended products:
+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+-------+------------+-------------------------+
|                                                                                                       Title                                                                                                        | Price | Avg Rating | Cosine Similarity Score |
+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+-------+------------+-------------------------+
|                                                                                    planters peanuts unsalted dry roasted 16 oz                                                          

### Evaluation

In [22]:
import pandas as pd
from sklearn.metrics import ndcg_score
import numpy as np

# Assuming get_recommendations_in_chunks is defined and returns a list of tuples (index, score)
product_index = 0
top_similar_products_bigram = get_recommendations_in_chunks(product_index, tfidf_matrix_reduced_bigram)

# Extract the indices and scores from the recommendations
top_indices = [index for index, score in top_similar_products_bigram]
top_scores = [score for index, score in top_similar_products_bigram]

# Assuming 'df' is your DataFrame and 'average_rating' is a column in your dataset
# Prepare relevance scores based on the average_rating of the recommended products
relevance_scores = df.loc[top_indices, 'average_rating'].tolist()

# Normalize relevance scores since NDCG in sklearn assumes relevance scores, not ratings directly
max_rating = max(relevance_scores)
normalized_relevance_scores = [score / max_rating for score in relevance_scores]

# Reshape for ndcg_score function
true_relevance = np.asarray([normalized_relevance_scores])
predicted_relevance = np.asarray([top_scores])

# Calculate NDCG score
ndcg = ndcg_score(true_relevance, predicted_relevance)
print("NDCG Score:", ndcg)


NDCG Score: 0.9807542734697822
